### PR TITLE
fix #4041, warn overriding Vue's internal methods

### LINKED ISSUE #4041: [Suggestion] Warn if user overrides Vue's internal methods
Currently, properties that start with `_` or `$` are not proxied on the Vue instance because they may conflict with Vue’s internal. The same doesn't apply for _methods_, leading to problematic cases like [this one](https://jsfiddle.net/phanan/ruvft5cy/1/), where users may (accidentally) define methods which override Vue's internal API. 

Suggestion: Though this arguably belongs in userland, it might be helpful of Vue to introduce a console warning (`Warn: You're overriding Vue's internal method $emit. Be aware of misbehaviors.`) for such cases.

### PR DESCRIPTION
**Pull request by HerringtonDarkholme**

### CODE REVIEW

File: src/core/instance/state.js [hunk 1]
```diff
@@ -143,12 +145,17 @@ function initMethods (vm: Component) {
...
-        warn(
+      if (process.env.NODE_ENV !== 'production') {
+        methods[key] == null && warn(
           `method "${key}" has an undefined value in the component definition. ` +
           `Did you reference the function correctly?`,
           vm
         )
+        hasOwn(BuiltinVue.prototype, key) && warn(
+          `You're overriding Vue's internal method "${key}". ` +
+          `Beware of misbehaviors.`,
```
Comment (yyx990803) [inline src/core/instance/state.js]: Let rephrase this to "Avoid overriding Vue's internaml method ..." and can remove the second line.

**Comment by phanan** [comment]
@yyx990803 As discussed on Slack, this commit only solves half of the problem, as it doesn't cover the _properties_, which can be more tricky.

File: src/core/instance/state.js [hunk 2]
```diff
@@ -143,12 +145,16 @@ function initMethods (vm: Component) {
...
       vm[key] = methods[key] == null ? noop : bind(methods[key], vm)
-      if (process.env.NODE_ENV !== 'production' && methods[key] == null) {
-        warn(
+      if (process.env.NODE_ENV !== 'production') {
+        methods[key] == null && warn(
           `method "${key}" has an undefined value in the component definition. ` +
           `Did you reference the function correctly?`,
           vm
         )
+        hasOwn(BuiltinVue.prototype, key) && warn(
```
Comment (phanan) [inline src/core/instance/state.js]: Can't we just use `vm` instead of importing Vue here?

Comment (HerringtonDarkholme) [inline src/core/instance/state.js]: Hi, I also considered that approach when coding this. But if we use `vm` or `vm.constructor.prototype`, it will break constructor style instance because `$emit` or so isn't their own properties.

**Comment by phanan** [comment]
Doesn't a simple `key in vm` work just fine?

**Comment by HerringtonDarkholme** [comment]
I'm sorry that this change introduced more problems than it resolved.
As @phanan pointed out, only `method` is warned against. Also https://github.com/vuejs/vue/issues/4142 is another problem caused by this.

I have reconsidered this. We want to only warn against overriding built-in method. And using `key in vm` to detect that effectively means:

* `vm` only has properties inherited from Vue.prototype. This is guaranteed by Vue's core code, where extending is done by option merging.
* Vue.prototype does not have user defined method. This is not controlled by core code, as #4142 showed.

@phanan Could you tell me the tricky part of properties? I'm not on slack.

**Comment by phanan** [comment]
@HerringtonDarkholme: @yyx990803 should have a better answer because he knows best about the framework's internal obviously, but from what I can tell, not all properties have been registered by the time methods are collected (examples are `$el` and `$slots`). This would make a check during `initMethods()` incomplete and was what keeping me from sending a PR over.

**Comment by HerringtonDarkholme** [comment]
Yes, I just tried to fix this again and found warning is particularly tricky thing. More importantly, a `_` and `$` prefix have been strong indicators to users. The marginal benefit for this pr doesn't worth...

**Comment by yyx990803** [comment]
I'll just revert this for now - I think this is a very rare case anyway. Or, we can simply change it to warn against any method that starts with `$` or `_`, since this is conventionally reserved for internal and plugin methods/properties.

**Comment by phanan** [comment]
Agreed. This "improvement" has only been costing us time with no particular benefits to start with :)
